serverless/virus-scanner/serverless.yml

```diff
@@ -0,0 +1,51 @@
+service: virus-scanner
+frameworkVersion: '3'
+useDotenv: true
+custom:
+  currentStage: ${opt:stage}
+  serverlessIfElse:
+    - If: '"${opt:stage}" == "production"'
+      Set:
+        - provider.provisionedConcurrency: 5
+      ElseSet:
+        - provider.provisionedConcurrency: 1
+provider:
+  name: aws
+  region: ap-southeast-1
+  stage: ${opt:stage}
+  provisionedConcurrency: 0
+  iam:
+    role:
+      statements:
+        # Allow functions to read/write objects in a bucket
+        - Effect: Allow
+          Action:
+            - 's3:GetObject'
+            - 's3:PutObject'
+          Resource:
+            - 'arn:aws:s3:::${env:VIRUS_SCANNER_QUARANTINE_S3_BUCKET}/*'
```

Comment:
Not sure if this is possible, but would be great to have this limited further to `'arn:aws:s3:::${env:VIRUS_SCANNER_QUARANTINE_S3_BUCKET}/${env:FILE_UUID}'` so that if for some reason there is a compromise at the lambda level, it should not have access to every other file in the clean/quarantine buckets.

Reply:
i spent some time investigating this:

- i don’t think it is possible to define the perms in advance at build/deployment time (viz serverless.yml) since the filenames are randomised

- at execution time, it should be possible for the ec2 instance to update the lambda function's role privileges to scope to the specific file being scanned, before executing the function. but i am slightly concerned that if we expect a high level of concurrency, there might be a race condition at some point, where:
- lambda role is updated by ec2 instance A to scope to instance A's file of interest
- lambda role is updated by ec2 instance B to scope to instance B's file of interest
- ec2 instance A calls lambda execution, which fails because it does not have the right permissions

- the risk is somewhat reduced now that we have enforced deletion of quarantine bucket file upon scan completion, and at the application level, we will delete the clean bucket file upon successful retrieval. finally, we have lifecycle rules as a catchall to delete files in s3 bucket after preset time

- The above considered, I'm not too inclined to implement further access control on the quarantine bucket at execution time

@spaceraccoon for your further thoughts pls, thanks very much for reviewing this!

Reply:
as spoken, this is resolved through the use of uuid filenames for defence in depth